File: cube_puzzle.scala

package org.danob.puzzle

import collection.mutable
import collection.immutable

sealed abstract class Axis
object Axis {
  case object X extends Axis
  case object Y extends Axis
  case object Z extends Axis
  val All = List(X, Y, Z)
}

sealed abstract class Direction(val sign:Int)
object Direction {
  case object Pos extends Direction(1) {
    override val toString = "+"
  }

  case object Neg extends Direction(-1) {
    override val toString = "-"
  }

  val All = List(Pos, Neg)
}

case class Move(axis:Axis, direction:Direction) {
  override val toString = direction.toString + axis.toString
}

object Move {
  val All:Iterable[Move] = Axis.All flatMap { axis ⇒
    Direction.All map { direction ⇒
      new Move(axis, direction)
    }
  }
}

case class Voxel(x:Int, y:Int, z:Int) {
  def apply(axis:Axis) = axis match {
    case Axis.X ⇒ x
    case Axis.Y ⇒ y
    case Axis.Z ⇒ z
  }

  def offset(axis:Axis, amount:Int) = axis match {
    case Axis.X ⇒ new Voxel(x+amount, y, z)
    case Axis.Y ⇒ new Voxel(x, y+amount, z)
    case Axis.Z ⇒ new Voxel(x, y, z+amount)
  }
}
case class SolvedException(state:PuzzleState) extends Exception
case class MoveNotAllowedException() extends Exception

class PuzzleState(
  val chainRemaining:Vector[Int],
  val voxelsOccupied:Set[Voxel],
  val moves:Vector[Move],
  val lastVoxel:Voxel,
  val ranges:immutable.Map[Axis, Tuple2[Int, Int]],
  val sharedSolveState:mutable.Map[String, Int]) {

  def with_move(move:Move):Option[PuzzleState] = {
    if (move.axis == moves.last.axis) {
      // cannot move same axis twice in a row
      return None
    }

    val next_chain_length = this.chainRemaining.head

    val new_voxels = new mutable.ArrayBuffer[Voxel]
    try {
      (1 until next_chain_length) foreach { s ⇒
        val new_voxel = lastVoxel.offset(move.axis, s * move.direction.sign)

        if (new_voxel(move.axis) < ranges(move.axis)._2 - 3) {
          // outside of the 4x4 cube, too negative
          throw new MoveNotAllowedException()
        }

        if (new_voxel(move.axis) > ranges(move.axis)._1 + 3) {
          // outside of the 4x4 cube, too positive
          throw new MoveNotAllowedException()
        }

        if (voxelsOccupied.contains(new_voxel)) {
          throw new MoveNotAllowedException()
        }

        // ok this voxel is good
        new_voxels.append(new_voxel)
      }
    } catch {
      case MoveNotAllowedException() ⇒ return None
    }

    val oldRange = this.ranges(move.axis)
    val newCoordinate = new_voxels.last(move.axis)
    val newRange = (
      if (newCoordinate < oldRange._1) newCoordinate else oldRange._1,
      if (newCoordinate > oldRange._2) newCoordinate else oldRange._2
    )

    Some(new PuzzleState(
      this.chainRemaining.tail,
      this.voxelsOccupied ++ new_voxels,
      moves :+ move,
      new_voxels.last,
      ranges + ((move.axis, newRange)),
      this.sharedSolveState
    ))
  }

  def solve():Option[PuzzleState] = {
    if (chainRemaining.isEmpty) {
      return Some(this)
    }

    try {
      Move.All foreach { move ⇒
        this.with_move(move) foreach { position ⇒
          sharedSolveState("positions") += 1
          if (sharedSolveState("positions") % 100000 == 0) {
            val p = sharedSolveState("positions")
            val d = sharedSolveState("depth")
            println(s"positions seen: $p (depth $d)")
          }

          sharedSolveState("depth") += 1
          position.solve() foreach { s:PuzzleState ⇒
            throw new SolvedException(s)
          }
          sharedSolveState("depth") -= 1
        }
      }
    } catch {
      case SolvedException(state) ⇒ return Some(state)
    }

    None
  }

  override def toString = {
    moves.tail mkString "\n"
  }
}

object PuzzleState {
  val startingChain = Vector(
    3, 2, 3, 2, 2, 4, 2, 3, 2, 3, 2, 3, 2, 2, 2, 2, 2, 2, 2, 2, 3, 3,
    2, 2, 2, 2, 2, 3, 4, 2, 2, 2, 4, 2, 3, 2, 2, 2, 2, 2, 2, 2, 2, 2,
    4, 2)

  val startingVoxel = Voxel(0, 0, 0)
  val startingVoxelsOccupied = Set(startingVoxel)
  val startingRanges:Map[Axis, Tuple2[Int, Int]] = Map(
    Axis.X → (0, 0),
    Axis.Y → (0, 0),
    Axis.Z → (0, 0))

  def apply() = new PuzzleState(
    startingChain,
    startingVoxelsOccupied,
    Vector[Move](new Move(Axis.Z, Direction.Pos)),
    startingVoxel,
    startingRanges,
    mutable.Map("positions" → 0, "depth" → 0)
  )
}

object Main extends App {
  val s = PuzzleState()
  val solution = s.solve()
  solution match {
    case Some(s) ⇒ println("solution: " + solution.toString)
    case None ⇒ println("no solution found!")
  }
}

File: cube_puzzle.py
#!/usr/bin/env python

# length: 46
STARTING_CHAIN = [
    2, 4, 2, 2, 2, 2, 2, 2, 2, 2,
    2, 3, 2, 4, 2, 2, 2, 4, 3, 2,
    2, 2, 2, 2, 3, 3, 2, 2, 2, 2,
    2, 2, 2, 2, 3, 2, 3, 2, 3, 2,
    4, 2, 2, 3, 2, 3
]

AXIS_X = 0
AXIS_Y = 1
AXIS_Z = 2


class PuzzleState:
    def initialize(self):
        self.chain_remaining = STARTING_CHAIN
        self.voxels_occupied = set([(0, 0, 0)])
        self.moves = [(AXIS_Z, 1)]
        self.current_voxel = (0, 0, 0)
        self.ranges = [[0, 0], [0, 0], [0, 0]]
        self.global_solve_state = {
            'depth': 0,
            'positions': 0
        }

    def copy(self):
        other = PuzzleState()
        other.chain_remaining = list(self.chain_remaining)
        other.voxels_occupied = set(self.voxels_occupied)
        other.moves = list(self.moves)
        other.current_voxel = self.current_voxel
        other.ranges = [list(x) for x in self.ranges]
        other.global_solve_state = self.global_solve_state
        return other

    def create_from_move(self, move):
        axis, sign = move

        if axis == self.moves[-1][0]:
            # cannot move same axis twice in a row
            return None

        result = self.copy()

        next_chain_length = result.chain_remaining[-1]

        # note that index zero is already placed; we start at 1.
        for s in range(1, next_chain_length):
            new_voxel = list(self.current_voxel)
            new_voxel[axis] += s * sign

            if new_voxel[axis] < result.ranges[axis][1] - 3:
                # outside of the 4x4 cube, too negative
                return None

            if new_voxel[axis] > result.ranges[axis][0] + 3:
                # outside of the 4x4 cube, too positive
                return None

            new_voxel = tuple(new_voxel)
            if new_voxel in result.voxels_occupied:
                # this voxel is occupied
                return None

            # ok this voxel is good
            result.voxels_occupied.add(new_voxel)
            if new_voxel[axis] < result.ranges[axis][0]:
                result.ranges[axis][0] = new_voxel[axis]
            if new_voxel[axis] > result.ranges[axis][1]:
                result.ranges[axis][1] = new_voxel[axis]
            result.current_voxel = new_voxel

        result.chain_remaining = result.chain_remaining[:-1]
        result.moves.append(move)

        return result

    def solve(self):
        if not self.chain_remaining:
            return self

        for axis in range(3):
            for sign in [1, -1]:
                new_cube = self.create_from_move((axis, sign))
                if new_cube:
                    self.global_solve_state['positions'] += 1
                    if self.global_solve_state['positions'] % 100000 == 0:
                        print('positions seen: {} (depth {})'.format(
                            self.global_solve_state['positions'],
                            self.global_solve_state['depth']))

                    self.global_solve_state['depth'] += 1
                    solution = new_cube.solve()
                    self.global_solve_state['depth'] -= 1

                    if solution:
                        return solution

        # no solution found from this spot
        return None

    def __repr__(self):
        axis_names = ['x', 'y', 'z']
        return '\n'.join([
            '{sign}{axis} ({length})'.format(
                sign='+' if sign > 0 else '-',
                axis=axis_names[axis],
                length=length)
            for (axis, sign), length
            in zip(self.moves[1:], STARTING_CHAIN[::-1])
        ])


def main():
    s = PuzzleState()
    s.initialize()
    s = s.solve()
    print('solution:\n{}'.format(s))


if __name__ == '__main__':
    main()
